### ISSUE TITLE
Does not support tasks_iam_role_statements through the single module?

### ISSUE THREAD

**Issue by dhutsj** [terraform-aws-modules/terraform-aws-ecs#114]
## Is your request related to a new offering from AWS?
Is this functionality available in the [AWS provider for Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs)? See [CHANGELOG.md](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md), too.
  - No 🛑: please wait to file a request until the functionality is avaialble in the AWS provider
  - Yes ✅: please list the AWS provider version which introduced this functionality

## Is your request related to a problem? Please describe.

As the README, I want to create resources through a single module that creates both the cluster and service resources. But by this single module, it only supports `task_exec_iam_statements` for task exec role, but does not support `tasks_iam_role_statements` for task role.

**Comment by bryantbiggs** [comment]
it does https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/2604124d05974c2ee47ff7194d62d55ac425a3cb/main.tf#L148

Are there any issues you are seeing with this?

**Comment by dhutsj** [comment]
> it does
> 
> https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/2604124d05974c2ee47ff7194d62d55ac425a3cb/main.tf#L148
> 
> Are there any issues you are seeing with this?

Yes, here is the code I use, then it pop up the below error.
```shell
module "ecs" {
  source  = "terraform-aws-modules/ecs/aws"
  version = "5.2.2"
  ...
  create_tasks_iam_role = true
  tasks_iam_role_statements = something
  }
```
The error message is below.
```shell
Error: Unsupported argument
│ 
│   on ecs.tf line 78, in module "ecs":
│   78:   create_tasks_iam_role = true
│ 
│ An argument named "create_tasks_iam_role" is not expected here.
╵
╷
│ Error: Unsupported argument
│ 
│   on ecs.tf line 79, in module "ecs":
│   79:   tasks_iam_role_statements = [
│ 
│ An argument named "tasks_iam_role_statements" is not expected here.
```
Thank you so much for looking at this.

**Comment by bryantbiggs** [comment]
Can you provide a proper reproduction

**Comment by dhutsj** [comment]
> Can you provide a proper reproduction

Yes, I can paste my code here.
```shell
module "ecs" {
  source  = "terraform-aws-modules/ecs/aws"
  version = "5.2.2"

  cluster_name = "xxx"

  cluster_configuration = {
    execute_command_configuration = {
      logging = "OVERRIDE"
      log_configuration = {
        cloud_watch_log_group_name = "xxx"
      }
    }
  }

  fargate_capacity_providers = {
    FARGATE = {
      default_capacity_provider_strategy = {
        weight = 100
      }
    }
  }

  services = {
    xxx = {
      cpu    = 4096
      memory = 4096
      volume = {
        my-vol = {
          efs_volume_configuration = {
            file_system_id     = module.efs.id
            root_directory     = ""
            transit_encryption = "DISABLED"
          }
        }
      }

      # Container definition(s)
      container_definitions = {

        xxx = {
          cpu       = 4096
          memory    = 4096
          essential = true
          image     = "placeholder"
          mount_points = [
            {
              sourceVolume  = "my-vol",
              containerPath = "/opt/backup"
            }
          ]
        }
      }

      subnet_ids = concat(data.aws_subnets.private2a.ids, data.aws_subnets.private2b.ids, data.aws_subnets.private2c.ids)
      security_group_rules = {
        egress_all = {
          type        = "egress"
          from_port   = 0
          to_port     = 0
          protocol    = "-1"
          cidr_blocks = ["0.0.0.0/0"]
        }
      }
    }
  }

  create_task_exec_iam_role = true
  create_task_exec_policy   = true
  task_exec_iam_statements = [
    {
      actions   = ["s3:PutObject"]
      effect    = "Allow"
      resources = ["arn:aws:s3:::xxx"]
      sid       = "S3"
    }
  ]

  create_tasks_iam_role = true
  tasks_iam_role_statements = something

  tags = local.common_tags
}
```
Plan for this code, it pops up the error `An argument named "create_tasks_iam_role" is not expected here.` and `An argument named "tasks_iam_role_statements" is not expected here.`

Reproduce everytime.

**Comment by dhutsj** [comment]
Hello @bryantbiggs , may I know do you get any chance to reproduce the issue or is it anything wrong about my usage of the module?
Thank you so much in advance.

**Comment by bryantbiggs** [comment]
tasks role is specific to the service/task definition so it goes inside the service definition. also, I see you have S3 permissions on the task exec role - this role is used during task startup. so if you are pulling images or getting secrets and populating the task definition, it needs permission to do those things. the AWS permissions required by your task, those will go in the tasks role

Here is an updated copy of the code you provided that should do what you are looking for (bar from the correct permissions which you will need to amend):

```hcl
module "ecs" {
  source  = "terraform-aws-modules/ecs/aws"
  version = "5.2.2"

  cluster_name = "xxx"

  cluster_configuration = {
    execute_command_configuration = {
      logging = "OVERRIDE"
      log_configuration = {
        cloud_watch_log_group_name = "xxx"
      }
    }
  }

  fargate_capacity_providers = {
    FARGATE = {
      default_capacity_provider_strategy = {
        weight = 100
      }
    }
  }

  services = {
    xxx = {
      cpu    = 4096
      memory = 4096
      volume = {
        my-vol = {
          efs_volume_configuration = {
            file_system_id     = module.efs.id
            root_directory     = ""
            transit_encryption = "DISABLED"
          }
        }
      }

      # Container definition(s)
      container_definitions = {

        xxx = {
          cpu       = 4096
          memory    = 4096
          essential = true
          image     = "placeholder"
          mount_points = [
            {
              sourceVolume  = "my-vol",
              containerPath = "/opt/backup"
            }
          ]
        }
      }

      subnet_ids = concat(data.aws_subnets.private2a.ids, data.aws_subnets.private2b.ids, data.aws_subnets.private2c.ids)
      security_group_rules = {
        egress_all = {
          type        = "egress"
          from_port   = 0
          to_port     = 0
          protocol    = "-1"
          cidr_blocks = ["0.0.0.0/0"]
        }
      }

      create_tasks_iam_role = true
      tasks_iam_role_statements = [
        {
          actions   = ["s3:PutObject"]
          effect    = "Allow"
          resources = ["arn:aws:s3:::xxx"]
          sid       = "S3"
        }
      ]
    }
  }

  tags = local.common_tags
}
```

**Comment by dhutsj** [comment]
> tasks role is specific to the service/task definition so it goes inside the service definition. also, I see you have S3 permissions on the task exec role - this role is used during task startup. so if you are pulling images or getting secrets and populating the task definition, it needs permission to do those things. the AWS permissions required by your task, those will go in the tasks role
> 
> Here is an updated copy of the code you provided that should do what you are looking for (bar from the correct permissions which you will need to amend):
> 
> ```terraform
> module "ecs" {
>   source  = "terraform-aws-modules/ecs/aws"
>   version = "5.2.2"
> 
>   cluster_name = "xxx"
> 
>   cluster_configuration = {
>     execute_command_configuration = {
>       logging = "OVERRIDE"
>       log_configuration = {
>         cloud_watch_log_group_name = "xxx"
>       }
>     }
>   }
> 
>   fargate_capacity_providers = {
>     FARGATE = {
>       default_capacity_provider_strategy = {
>         weight = 100
>       }
>     }
>   }
> 
>   services = {
>     xxx = {
>       cpu    = 4096
>       memory = 4096
>       volume = {
>         my-vol = {
>           efs_volume_configuration = {
>             file_system_id     = module.efs.id
>             root_directory     = ""
>             transit_encryption = "DISABLED"
>           }
>         }
>       }
> 
>       # Container definition(s)
>       container_definitions = {
> 
>         xxx = {
>           cpu       = 4096
>           memory    = 4096
>           essential = true
>           image     = "placeholder"
>           mount_points = [
>             {
>               sourceVolume  = "my-vol",
>               containerPath = "/opt/backup"
>             }
>           ]
>         }
>       }
> 
>       subnet_ids = concat(data.aws_subnets.private2a.ids, data.aws_subnets.private2b.ids, data.aws_subnets.private2c.ids)
>       security_group_rules = {
>         egress_all = {
>           type        = "egress"
>           from_port   = 0
>           to_port     = 0
>           protocol    = "-1"
>           cidr_blocks = ["0.0.0.0/0"]
>         }
>       }
> 
>       create_tasks_iam_role = true
>       tasks_iam_role_statements = [
>         {
>           actions   = ["s3:PutObject"]
>           effect    = "Allow"
>           resources = ["arn:aws:s3:::xxx"]
>           sid       = "S3"
>         }
>       ]
>     }
>   }
> 
>   tags = local.common_tags
> }
> ```

I tried the above code. Same error.
```shell
Error: Unsupported argument
│ 
│   on ecs.tf line 78, in module "ecs":
│   78:   create_tasks_iam_role = true
│ 
│ An argument named "create_tasks_iam_role" is not expected here.
╵
╷
│ Error: Unsupported argument
│ 
│   on ecs.tf line 79, in module "ecs":
│   79:   tasks_iam_role_statements = [
│ 
│ An argument named "tasks_iam_role_statements" is not expected here.
```
So I think as long as you use the single module which according to the README, it fails to pass `tasks_iam_role`.
```shell
module "ecs" {
  source  = "terraform-aws-modules/ecs/aws"
```
I think it work for use it like below.
```shell
module "ecs_service" {
  source  = "terraform-aws-modules/ecs/aws//modules/service"
```
But the thing is it should work through the wrapper cause the single module contain cluster service and task, all in one. That's the reason why I raised this issue call it `Does not support tasks_iam_role_statements through the single module?`.
Thank you.

**Comment by antonbabenko** [comment]
This issue has been resolved in version 5.3.0 :tada:
